app/cdap/components/NamespaceAdmin/SourceControlManagement/SourceControlManagementModal.tsx

```diff
@@ -0,0 +1,291 @@
+}
+
+interface ISourceControlManagementFormProps {
+  initialSourceControlManagementState?: ISourceControlManagement | null;
+  onSubmit: (formState: ISourceControlManagement) => void;
+  toggleModal: () => void;
+}
+
+const SourceControlManagementForm = ({
+  initialSourceControlManagementState,
+  onSubmit,
+  toggleModal,
+}: ISourceControlManagementFormProps) => {
+  const [formState, dispatch] = useReducer(
+    sourceControlManagementFormReducer,
+    initialSourceControlManagementState
+  );
+
+  const [linkError, setLinkError] = useState(false);
+  const [tokenNameError, setTokenNameError] = useState(false);
+  const [tokenError, setTokenError] = useState(false);
+  const [missingFields, setMissingFields] = useState(0);
+  const [showValidateMessage, setShowValidateMessage] = useState(false);
+
+  useEffect(() => {
+    if (!formState || !formState.auth || !formState.auth.type) {
+      dispatch({
+        type: 'RESET',
+      });
+    }
+  }, []);
+
+  const handleValueChange = (e, type) => {
+    e.preventDefault();
+    dispatch({
+      type,
+      value: e.target.value,
+    });
+  };
+
+  const validateForm = () => {
+    let missingFieldsCount = 0;
+    let validateSuccess = true;
+    if (!formState.link) {
+      setLinkError(true);
+      missingFieldsCount++;
+      validateSuccess = false;
+    }
+    if (!formState.auth.token) {
+      setTokenError(true);
+      missingFieldsCount++;
+      validateSuccess = false;
+    }
+    if (!formState.auth.tokenName) {
+      setTokenNameError(true);
+      missingFieldsCount++;
+      validateSuccess = false;
+    }
+    setMissingFields(missingFieldsCount);
+    setShowValidateMessage(true);
+    return validateSuccess;
+  };
+
+  const handleSubmit = () => {
+    if (!validateForm()) {
+      return;
+    }
+    onSubmit(formState);
+  };
+  return (
+    <ThemeProvider theme={AsteriskTheme}>
+      <StyledForm>
+        <BoldHeader>{T.translate(`${PREFIX}.basicHeader`)}</BoldHeader>
+        <FormGroup>
+          <Select
+            label={T.translate(`${PREFIX}.provider.label`)}
+            value={formState?.provider ? formState?.provider : providers.github}
+            onChange={(e) => {
+              handleValueChange(e, 'SET_PROVIDER');
+            }}
+            style={{ width: '100%' }}
+          >
+            {Object.values(providers).map((provider) => (
+              <MenuItem value={provider}>{provider}</MenuItem>
+            ))}
+          </Select>
+          <FormHelperText>{T.translate(`${PREFIX}.provider.helperText`)}</FormHelperText>
+          <StyledTextField
+            label={T.translate(`${PREFIX}.repoUrl.label`)}
+            value={formState?.link}
+            helperText={T.translate(`${PREFIX}.repoUrl.helperText`)}
+            onChange={(e) => {
+              handleValueChange(e, 'SET_REPOSITORY_URL');
+              setLinkError(!e.target.value);
+            }}
+            required
+            error={linkError}
```

Comment:
can't this just be formState?.link === '' or something like that?

Reply:
changed